modules/utxo-lib/src/bitgo/UtxoPsbt.ts

```diff
@@ -0,0 +1,319 @@
+    }
+    const input = checkForInput(this.data.inputs, inputIndex);
+    if (!input.tapBip32Derivation || input.tapBip32Derivation.length === 0) {
+      throw new Error('Need tapBip32Derivation to sign Taproot with HD');
+    }
+    const myDerivations = input.tapBip32Derivation
+      .map((bipDv) => {
+        if (bipDv.masterFingerprint.equals(hdKeyPair.fingerprint)) {
+          return bipDv;
+        }
+      })
+      .filter((v) => !!v) as TapBip32Derivation[];
+    if (myDerivations.length === 0) {
+      throw new Error('Need one tapBip32Derivation masterFingerprint to match the HDSigner fingerprint');
+    }
+    const signers: TaprootSigner[] = myDerivations.map((bipDv) => {
+      const node = hdKeyPair.derivePath(bipDv.path);
+      if (!bipDv.pubkey.equals(node.publicKey.slice(1))) {
+        throw new Error('pubkey did not match tapBip32Derivation');
+      }
+      return { signer: node, leafHashes: bipDv.leafHashes };
+    });
+    signers.forEach(({ signer, leafHashes }) => this.signTaprootInput(inputIndex, signer, leafHashes, sighashTypes));
+    return this;
+  }
+
+  signTaprootInput(
+    inputIndex: number,
+    signer: SchnorrSigner,
+    leafHashes: Buffer[],
+    sighashTypes: number[] = [Transaction.SIGHASH_DEFAULT, Transaction.SIGHASH_ALL]
+  ): this {
+    const input = checkForInput(this.data.inputs, inputIndex);
+    // Figure out if this is script path or not, if not, tweak the private key
+    if (!input.tapLeafScript?.length) {
+      // See BitGo/BitGoJS/modules/utxo_lib/src/transaction_builder.ts:trySign for how to support it.
+      throw new Error('Taproot key path signing is not supported.');
+    }
+    if (input.tapLeafScript.length !== 1) {
+      throw new Error('Only one leaf script supported for signing');
+    }
+    const tapLeafScript = input.tapLeafScript[0];
+    const parsedControlBlock = taproot.parseControlBlock(eccLib, tapLeafScript.controlBlock);
+    const { leafVersion } = parsedControlBlock;
+    if (leafVersion !== tapLeafScript.leafVersion) {
+      throw new Error('Tap script leaf version mismatch with control block');
+    }
+    const leafHash = taproot.getTapleafHash(eccLib, parsedControlBlock, tapLeafScript.script);
+    if (!leafHashes.find((l) => l.equals(leafHash))) {
+      throw new Error(`Signer cannot sign for leaf hash ${leafHash.toString('hex')}`);
+    }
+    const { hash, sighashType } = this.getTaprootHashForSig(inputIndex, false, sighashTypes, leafHash);
+    let signature = signer.signSchnorr(hash);
+    if (sighashType !== Transaction.SIGHASH_DEFAULT) {
+      signature = Buffer.concat([signature, Buffer.of(sighashType)]);
+    }
+    this.data.updateInput(inputIndex, {
+      tapScriptSig: [
+        {
+          pubkey: signer.publicKey.slice(1),
+          signature,
+          leafHash,
+        },
+      ],
+    });
+    return this;
+  }
+
+  private getTaprootHashForSig(
+    inputIndex: number,
+    forValidate: boolean,
```

Comment:
unused - what was the intended purpose?

Reply:
Exists because it exists on the upstream equivalent method. Would have to go to the upstream `getHashForSig` and read the code to remember why it exists there.